src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/ErrorExtensions.cs

```diff
@@ -0,0 +1,122 @@
+        /// <returns></returns>
+        public static SNIErrorDetails ToSniErrorDetails(this SNIError sniError)
+            => new SNIErrorDetails
+            {
+                sniErrorNumber = sniError.sniError,
+                errorMessage = sniError.errorMessage,
+                nativeError = sniError.nativeError,
+                provider = (int)sniError.provider,
+                lineNumber = sniError.lineNumber,
+                function = sniError.function,
+                exception = sniError.exception
+            };
+
+        public static SqlError ToSqlError(this SNIError sniError, SniContext sniContext, ServerInfo info)
+        {
+            SNIErrorDetails details = sniError.ToSniErrorDetails();
+
+            if (details.sniErrorNumber != 0)
+            {
+                // handle special SNI error codes that are converted into exception which is not a SqlException.
+                switch (details.sniErrorNumber)
+                {
+                    case (int)SNINativeMethodWrapper.SniSpecialErrors.MultiSubnetFailoverWithMoreThan64IPs:
+                        // Connecting with the MultiSubnetFailover connection option to a SQL Server instance configured with more than 64 IP addresses is not supported.
+                        SqlClientEventSource.Log.TryAdvancedTraceEvent("<sc.TdsParser.ProcessSNIError|ERR|ADV> Connecting with the MultiSubnetFailover connection option to a SQL Server instance configured with more than 64 IP addresses is not supported.");
+                        throw SQL.MultiSubnetFailoverWithMoreThan64IPs();
+
+                    case (int)SNINativeMethodWrapper.SniSpecialErrors.MultiSubnetFailoverWithInstanceSpecified:
+                        // Connecting to a named SQL Server instance using the MultiSubnetFailover connection option is not supported.
+                        SqlClientEventSource.Log.TryAdvancedTraceEvent("<sc.TdsParser.ProcessSNIError|ERR|ADV> Connecting to a named SQL Server instance using the MultiSubnetFailover connection option is not supported.");
+                        throw SQL.MultiSubnetFailoverWithInstanceSpecified();
+
+                    case (int)SNINativeMethodWrapper.SniSpecialErrors.MultiSubnetFailoverWithNonTcpProtocol:
+                        // Connecting to a SQL Server instance using the MultiSubnetFailover connection option is only supported when using the TCP protocol.
+                        SqlClientEventSource.Log.TryAdvancedTraceEvent("<sc.TdsParser.ProcessSNIError|ERR|ADV> Connecting to a SQL Server instance using the MultiSubnetFailover connection option is only supported when using the TCP protocol.");
+                        throw SQL.MultiSubnetFailoverWithNonTcpProtocol();
+                        // continue building SqlError instance
+                }
+            }
+
+            SqlClientEventSource.Log.TryAdvancedTraceEvent("< sc.TdsParser.ProcessSNIError |ERR|ADV > Error message Detail: {0}", details.errorMessage);
```

Comment:
The spacing suggestion is fine. But we try to avoid formatting strings in logging parameters for perf. We want the string formatting to happen in the logging function only if logging is enabled.

Reply:
That makes sense, thanks. The main goal was to avoid any overhead in parsing a compile-time constant format string at runtime; that should still happen, but I've suggested it a layer too high. The spacing's a nit.